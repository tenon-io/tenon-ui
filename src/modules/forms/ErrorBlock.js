import React, { Component, forwardRef } from 'react';
import PropTypes from 'prop-types';
import Heading from '../layout/Heading';
import uuidv4 from 'uuid/v4';

/**
 * @component
 * Displays the error texts for the collection of form controls
 * and renders as <a> tags with href attributes navigating to
 * the invalid controls.
 *
 * @prop required {object} formControls
 *
 * @example
 * example
 * {
 *      formControls: {
 *          [controlName]: {
 *              [controlId]: string,
 *              [value]: string,
 *              [validity]: boolean,
 *              [errorText]: string
 *          },
 *          [validity]: boolean
 *      }
 * }
 *
 * @prop required {string} headingText - Text to show as
 *                  block heading.
 */
class ErrorBlock extends Component {
    static propTypes = {
        formControls: PropTypes.shape({
            name: PropTypes.shape({
                controlId: PropTypes.string,
                value: PropTypes.string,
                validity: PropTypes.bool,
                errorText: PropTypes.string
            })
        }).isRequired,
        headingText: PropTypes.string.isRequired
    };

    constructor(props) {
        super(props);

        this.headingId = uuidv4();
    }

    render() {
        const { formControls, headingText, forwardedRef } = this.props;
        const controlNameArray = Object.keys(formControls);
        return controlNameArray.length > 0 ? (
            <section
                className="error-block"
                tabIndex="-1"
                ref={forwardedRef}
                aria-labelledby={this.headingId}
            >
                <Heading.H id={this.headingId} className="block-heading">
                    <span className="icon" />
                    {headingText}
                </Heading.H>
                <ul>
                    {controlNameArray
                        .filter(control => !formControls[control].validity)
                        .map(control => (
                            <li key={formControls[control].controlId}>
                                <a href={`#${formControls[control].controlId}`}>
                                    {formControls[control].errorText}
                                </a>
                            </li>
                        ))}
                </ul>
            </section>
        ) : null;
    }
}

export default forwardRef(({ formControls, headingText }, ref) => (
    <ErrorBlock
        formControls={formControls}
        headingText={headingText}
        forwardedRef={ref}
    />
));
