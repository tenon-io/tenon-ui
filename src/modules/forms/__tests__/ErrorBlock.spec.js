jest.mock('uuid/v4');

import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import ErrorBlock from '../ErrorBlock';
import uuidv4 from 'uuid/v4';

describe('ErrorBlock', () => {
    afterEach(cleanup);

    beforeEach(() => {
        jest.clearAllMocks();
        uuidv4.mockReturnValueOnce('sectionId');
    });

    it('should render as an unordered list to track the number of items', () => {
        const mockControls = {
            username: {
                validity: false,
                controlId: 'usernameControl',
                errorText: 'User name is required.'
            }
        };
        const { container, getByText } = render(
            <ErrorBlock formControls={mockControls} headingText="Some errors" />
        );
        expect(getByText('User name is required.')).toHaveAttribute(
            'href',
            '#usernameControl'
        );
        expect(container).toMatchSnapshot();
    });

    it('should render the error text and a hash href for each link', () => {
        const mockControls = {
            password: {
                validity: false,
                controlId: 'passwordControl',
                errorText: 'Your password is too short.'
            },
            tfaToken: {
                validity: false,
                controlId: 'tfaControl',
                errorText: 'Please enter a TFA token.'
            }
        };
        const { getByText } = render(
            <ErrorBlock formControls={mockControls} headingText="Some errors" />
        );
        expect(getByText('Your password is too short.')).toHaveAttribute(
            'href',
            '#passwordControl'
        );
        expect(getByText('Please enter a TFA token.')).toHaveAttribute(
            'href',
            '#tfaControl'
        );
    });

    it('should only render links for invalid controls', () => {
        const mockControls = {
            username: {
                validity: true,
                controlId: 'usernameControl',
                errorText: ''
            },
            password: {
                validity: false,
                controlId: 'passwordControl',
                errorText: 'Your password is too short.'
            },
            tfaToken: {
                validity: false,
                controlId: 'tfaControl',
                errorText: 'Please enter a TFA token.'
            },
            somethingElse: {
                validity: true,
                controlId: 'somethinElseControl',
                errorText: ''
            }
        };
        const { container, getByText } = render(
            <ErrorBlock formControls={mockControls} headingText="Some errors" />
        );
        expect(getByText('Your password is too short.')).toBeInTheDocument();
        expect(getByText('Please enter a TFA token.')).toBeInTheDocument();
        const list = container.querySelector('ul');
        expect(list.querySelectorAll('a').length).toBe(2);
    });

    it('should not render for an empty object to avoid polluting the DOM with an empty ul', () => {
        const { container } = render(
            <div>
                <ErrorBlock formControls={{}} headingText="Some errors" />
            </div>
        );
        expect(container.firstChild).toBeEmpty();
    });
});
