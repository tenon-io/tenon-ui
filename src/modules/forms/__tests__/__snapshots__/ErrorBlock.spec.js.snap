// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBlock should render as an unordered list to track the number of items 1`] = `
<div>
  <section
    aria-labelledby="sectionId"
    class="error-block"
    tabindex="-1"
  >
    <h1
      class="block-heading"
      id="sectionId"
    >
      <span
        class="icon"
      />
      Some errors
    </h1>
    <ul>
      <li>
        <a
          href="#usernameControl"
        >
          User name is required.
        </a>
      </li>
    </ul>
  </section>
</div>
`;
